### PR TITLE
Add Simple Streaming implementation

### PR DESCRIPTION
**Pull request by Michicosun**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add streaming mode for every table engine via 2 processors in pipeline

example select:

```sql
set allow_experimental_streaming=1;
set allow_experimental_analyzer=1;

CREATE TABLE test (a String, b UInt64) ENGINE = MergeTree() ORDER BY a;

-- infinite query
SELECT a, b * b FROM test STREAM;

```

insert works as usual

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit e8e46df57bfe75a5a1615ea2b4db91216dde6cbc with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57830/e8e46df57bfe75a5a1615ea2b4db91216dde6cbc/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Review by al13n321 — CHANGES_REQUESTED** [review]
Looks really good overall. Here are some comments. The important ones are:
 * subscriptions potentially eating lots of memory in non-streaming queries (search for "non-streaming") - plz check if it's true and/or fix it,
 * don't rely on timing so much in the test.

File: src/Planner/CollectTableExpressionData.cpp [hunk 1]
```diff
@@ -192,13 +193,41 @@ void checkStorageSupportPrewhere(const QueryTreeNodePtr & table_expression)
     }
 }
 
+// checks that there are only user configured columns in requested set
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Planner/CollectTableExpressionData.cpp:196]: yes, at this point I tried to prohibit streaming queries with virtual columns, because by default for all tables such columns will be set in the TableSource, but in streaming queries, the blocks that the SubscriptionSource receives are user input blocks.

File: tests/queries/0_stateless/02952_simple_streaming_read.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    pkill -P $$
+    sleep 0.5
+}
+
+for engine_type in 'MergeTree() ORDER BY (a)' 'TinyLog' 'StripeLog' 'Memory';
+do
+   run_test "$engine_type"
+done
+
+
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply tests/queries/0_stateless/02952_simple_streaming_read.sh:58]: Thanks for this advice, by default many of this queries will not emit anything, but they will not stuck - at any time you can cancel the query. For now i can add some check that will fail queries with union or CreatingSet step from streaming source, but in the future it can be possible to replace such query plan steps with streaming analogs.

Second query for example is a good case for supportsStreaming IStorage method

[the remark this answers is not in the document]
Comment (al13n321) [reply tests/queries/0_stateless/02952_simple_streaming_read.sh:58]: Right, I guess it makes sense to just say that STREAM makes the query run forever, and there's no need to prevent the user from trying to use it where it doesn't make sense (e.g. in IN subquery - the user is maybe supposed to understand that the main query can't make progress until the IN subquery completes).

So the expected behavior for all of the above queries is to either stall or work. May make sense to smoke test it anyway, to make sure it doesn't crash, if it's easy to do (just start the query, sleep 0.5, kill it, repeat for a bunch of different queries?). For UNION ALL maybe also check that it actually works, i.e. sees inserts from both tables?

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 3]
```diff
@@ -8172,9 +8172,11 @@ StorageSnapshotPtr MergeTreeData::getStorageSnapshot(const StorageMetadataPtr &
 {
     auto snapshot_data = std::make_unique<SnapshotData>();
     ColumnsDescription object_columns_copy;
+    StreamSubscriptionPtr stream_subscription = nullptr;
 
     {
         auto lock = lockParts();
+        stream_subscription = subscribeForChanges();
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Storages/MergeTree/MergeTreeData.cpp:8179]: Generally storage will not have atomicity guarantees, but it can be implemented. It was just necessary to perform all operations with the subscription manager under the same lock as for the read/write pipelines.

For example StripeLog should create subscription in read method(because it has no snapshot realization) and with this subscription it starts pipeline, write on the other hand just push chunks to subscribers in the sink under the same lock. Serialization of {read, create subscription} = A and {write chunk to storage, push chunk to subscribers} = B will lead to one of executions: AB, BA. In first - read is already configured, so it will read from prepared snapshot(prepared in read) -> write to storage will not be visible, but will be visible subscribers chunk. In second case write will not see subscription for not started query -> chunk will be added to storage only.

SinkToSubscribers just naive implementation of this protocol compatible with any storage.

MergeTree does not have an atomic subscription at this stage of implementation. My idea was to change toSubscribersWrite for MergeTree - just do nothing in this method and push to subscribers chunks under the part-lock right in the MergeTreeData Transaction or under shared with it lock. 

This solution leads to 2 problems:
- write should always go to subscription manager, but it is a lightweight operation, because on write path only read lock is needed.
- write should create pulling pipeline from part and start reading from it - this can be a little expensive.

File: src/Processors/StreamingAdapter.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <Poco/Logger.h>
+
+#include <Processors/IProcessor.h>
+
+namespace DB
+{
+
+class StreamingAdapter final : public IProcessor
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Processors/StreamingAdapter.h:10]: I think yes, for now it can be replaced to ConcatProcessor, idea behind this class to make chunk reordering with respect to their cursors.

Proposed solution in issue operates with cursors - a map with stream-name -> offset.

Idea of how to manage storage cursors: If storage supports cursors, it should set up cursor for each chunk generated from source and in sink it also should set up cursor for a chunk before pushing it to subscribers. Cursor can be stored in ChunkInfo for example. When chunks are being consumed by StreamingAdapter it will sort them accordingly to their offsets inside this map. As far as I can tell, chunks at this part of pipeline could not contain multiple stream-name in cursor, because stream-name ~= partition-id. 

tldr: after this processor storage must be visible as FIFO Queue.

But this is just an idea.

File: tests/queries/0_stateless/02952_simple_streaming_read.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    # stop reading by killing client job
+    pkill -P $$
+    sleep 0.5
+}
+
+for engine_type in 'MergeTree() ORDER BY (a)' 'TinyLog' 'StripeLog' 'Memory';
+do
+   run_test "$engine_type"
+done
+
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply tests/queries/0_stateless/02952_simple_streaming_read.sh:57]: Thanks, I didn't take this case into account, perhaps the best thing to do if the source received a chunk with unexpected metadata is to terminate query with some kind of error.

File: src/Storages/Streaming/SubscriptionManager.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <Processors/Chunk.h>
+
+#include <Storages/Streaming/Subscription_fwd.h>
+
+namespace DB
+{
+
+// Structure for managing the subscriptions,
+// necessary for streaming requests to work
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Storages/Streaming/SubscriptionManager.h]: corrected the description

File: src/Common/ErrorCodes.cpp [hunk 7]
```diff
@@ -590,6 +590,7 @@
     M(708, ILLEGAL_STATISTIC) \
     M(709, CANNOT_GET_REPLICATED_DATABASE_SNAPSHOT) \
     \
+    M(998, FORBID_FOR_STREAMING_QUERIES) \
```
Comment (al13n321) [inline src/Common/ErrorCodes.cpp:593]: I'm not sure what the exact convention is here, but it sure looks like this error belongs one line higher (with value 710), next to all the normal errors. The errors below correspond to big basic components.

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (al13n321) [reply src/Storages/MergeTree/MergeTreeData.cpp:8179]: > write should create pulling pipeline from part and start reading from it - this can be a little expensive.

To get the block(s) to push to subscribers? Can't we use the block(s) that was used to write the part?

File: tests/queries/0_stateless/02952_simple_streaming_read.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (Michicosun) [reply tests/queries/0_stateless/02952_simple_streaming_read.sh]: added one more test to check the corner cases

File: tests/queries/0_stateless/02952_simple_streaming_read.sh [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "INSERT INTO t_streaming_test VALUES ('a', 100)"
+    sleep 0.5
+
+    # spawn stream
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT a, a || '-' || a, b * b FROM t_streaming_test STREAM" &
+    sleep 0.5
+
+    # insert some data into stream
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "INSERT INTO t_streaming_test (*) select number as a, number as b from numbers(7)"
+    sleep 0.5
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply tests/queries/0_stateless/02952_simple_streaming_read.sh]: thanks for the advice, I just reworked it a bit and put it in the bash test library.

File: tests/queries/0_stateless/02952_simple_streaming_read.sh [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    # clear table
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "TRUNCATE t_streaming_test"
+
+    # spawn 2 streams
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT a FROM t_streaming_test STREAM WHERE b % 2 == 0" &
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT a FROM t_streaming_test STREAM WHERE b % 2 == 1" &
+    sleep 0.5
+
+    # insert some data into stream
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "INSERT INTO t_streaming_test (*) select 'same-x-2' as a, number as b from numbers(2)"
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply tests/queries/0_stateless/02952_simple_streaming_read.sh]: added this case into test

File: src/Common/ErrorCodes.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Common/ErrorCodes.cpp]: thanks! changed to normal codes.

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Storages/MergeTree/MergeTreeData.cpp]: I double checked this, and no, the subscription is inside the storage snapshot, and it will only live at the  analyzer-planner stage, but it's still a good idea to propagate the flag to the getStorageSnapshot method.

please take a look at the structure StorageSnapshotSettings - added it to have another configurable place to create storage snapshots.

**Review by al13n321 — APPROVED** [review]
The changes look great! Sorry for the slow response. Please resolve the conflicts, update SettingsChangesHistory.h, optionally address the new nitpicks, and should be good to go.

File: src/Storages/IStorage.cpp [hunk 10]
```diff
@@ -311,6 +372,36 @@ void IStorage::restoreDataFromBackup(RestorerFromBackup & restorer, const String
                         getStorageID().getFullTableName(), data_path_in_backup);
 }
 
+StorageSnapshotPtr IStorage::getStorageSnapshot(const StorageMetadataPtr & metadata_snapshot, ContextPtr query_context) const
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Storages/IStorage.cpp:375]: actually it is because of clang-tidy warning: `Default arguments on virtual or override methods are prohibited`

File: src/Storages/MergeTree/MergeTreeSink.cpp [hunk 11]
```diff
@@ -188,6 +194,12 @@ void MergeTreeSink::finishDelayedChunk()
 
             added = storage.renameTempPartAndAdd(part, transaction, lock);
             transaction.commit(&lock);
+
+            /// if block is added: push current_block to all subscribers under the same lock as commit.
+            /// But there may be a race between insert and creating a new subscription, in which case
+            /// the block may already have been cleared, but this is normal, because this is concurrent operations.
```
[the remark this answers is not in the document]
Comment (Michicosun) [reply src/Storages/MergeTree/MergeTreeSink.cpp:200]: That is absolutely correct! The main idea that in case of concurrent insert (block P) and starting streaming query, it may not read the block P.

**Comment by Michicosun** [comment]
@al13n321 thank you for review!!

But I'm not recommend you to merge this pr, because:
- I have complete implementation of [streaming queries with cursors](https://github.com/ClickHouse/ClickHouse/pull/63312) in different pr and it has no conflicts.
- In this pr ReplicatedMergeTree is not supported and it cannot be, because streaming queries from replicated tables require cursors.

I see that issues are not critical, so maybe you can also check out final pr? [63312](https://github.com/ClickHouse/ClickHouse/pull/63312)

I just merged this pr and [queue mode](https://github.com/ClickHouse/ClickHouse/pull/59827) into single and above that implemented cursors. Because commits are the same, I do not close previous prs, to make easier review. 

Sorry about this move, but I needed to provide the final implementation of streaming queries read mechanics by the end of May (At least in the form of a pull request).
